src/main/java/io/nats/client/impl/MessageQueue.java

```diff
@@ -76,162 +68,104 @@ boolean isDraining() {
     }
 
     boolean isDrained() {
+        // poison pill is not included in the length count, or the size
         return this.running.get() == DRAINING && this.length() == 0;
     }
 
-    void signalOne() {
-        Thread t = waiters.poll();
-        if (t != null) {
-            LockSupport.unpark(t);
-        }
-    }
-
-    void signalIfNotEmpty() {
-        if (this.length.get() > 0) {
-            signalOne();
-        }
-    }
-
-    void signalAll() {
-        Thread t = waiters.poll();
-        while(t != null) {
-            LockSupport.unpark(t);
-            t = waiters.poll();
-        }
-    }
-
     void push(NatsMessage msg) {
-
         // If we aren't running, then we need to obey the filter lock
         // to avoid ordering problems
         if(!this.isRunning()) {
             this.filterLock.lock();
-            this.queue.add(msg);
+            this.put(msg);
             this.filterLock.unlock();
             this.sizeInBytes.getAndAdd(msg.getSizeInBytes());
             this.length.incrementAndGet();
-            signalOne();
             return;
         }
 
-        this.queue.add(msg);
+        this.put(msg);
         this.sizeInBytes.getAndAdd(msg.getSizeInBytes());
         this.length.incrementAndGet();
-        signalOne();
     }
 
-    NatsMessage waitForTimeout(Duration timeout) throws InterruptedException {
-        long timeoutNanos = (timeout != null) ? timeout.toNanos() : -1;
-        NatsMessage retVal = null;
-
-        if (timeoutNanos >= 0) {
-            Thread t = Thread.currentThread();
-            long start = System.nanoTime();
+    void poisonTheQueue() {
+        try {
+            this.queue.add(this.poisonPill);
+        } catch (IllegalStateException ie) { // queue was full, so we don't really need poison pill
+            // ok to ignore this
+        }
+    }
 
-            // Semi-spin for at most MAX_SPIN_TIME
-            if (timeoutNanos > MAX_SPIN_TIME) {
-                int count = 0;
-                while (this.isRunning() && (retVal = this.queue.poll()) == null && count < this.maxSpins) {
+    void put(NatsMessage msg) {
+        try {
+            this.queue.put(msg);
+        } catch (InterruptedException ie) {
+            // ok to ignore this
+        }
+    }
 
-                    if (this.isDraining()) {
-                        break;
-                    }
+    NatsMessage poll(Duration timeout) throws InterruptedException {
+        NatsMessage msg = null;
+        
+        if (timeout == null || this.isDraining()) { // try immediately
+            msg = this.queue.poll();
+        } else {
+            long nanos = timeout.toNanos();
 
-                    count++;
-                    LockSupport.parkNanos(SPIN_WAIT);
+            if (nanos != 0) {
+                msg = this.queue.poll(nanos, TimeUnit.NANOSECONDS);
+            } else {
+                while (this.isRunning()) {
+                    msg = this.queue.poll(100, TimeUnit.DAYS);

```

Comment:
So will this potentially block for 100 days?

Reply:
ultimately it is supposed to block forever, so anything i put here seems reasonable. The poison pill stops the wait if it has to.

Reply:
i don't want to spin too fast, so i figured bigger is better

Reply:
Maybe comment explaining so.

Reply:
will add comments throughout

Reply:
ok